news/views.py

```diff
@@ -599,3 +581,64 @@ def list_newsletters(request):
     active_newsletters = Newsletter.objects.filter(active=True)
     return render(request, 'news/newsletters.html',
                   {'newsletters': active_newsletters})
+
+
+def update_user_task(request, api_call_type, data=None, optin=True, sync=False):
+    """Call the update_user task async with the right parameters.
+
+    If sync==True, be sure to do the update synchronously and include the token
+    in the response.  Otherwise, basket has the option to do it all later
+    in the background.
+    """
+    data = data or request.POST.dict()
+
+    newsletters = data.get('newsletters', None)
+    if newsletters:
+        if api_call_type == SUBSCRIBE:
+            all_newsletters = newsletter_and_group_slugs()
+        else:
+            all_newsletters = newsletter_slugs()
+        for nl in [x.strip() for x in newsletters.split(',')]:
+            if nl not in all_newsletters:
+                return HttpResponseJSON({
+                    'status': 'error',
+                    'desc': 'invalid newsletter',
+                    'code': errors.BASKET_INVALID_NEWSLETTER,
+                }, 400)
+
+    if 'lang' in data and not language_code_is_valid(data['lang']):
+        return HttpResponseJSON({
+            'status': 'error',
+            'desc': 'invalid language',
+            'code': errors.BASKET_INVALID_LANGUAGE,
+        }, 400)
+
+    email = data.get('email')
+    token = data.get('token')
+    if not (email or token):
+        return HttpResponseJSON({
+            'status': 'error',
+            'desc': MSG_EMAIL_OR_TOKEN_REQUIRED,
+            'code': errors.BASKET_USAGE_ERROR,
+        }, 400)
+
+    if sync:
+        # We need a token for this user. If we don't have a Subscriber
+        # object for them already, we'll need to find or make one,
+        # checking ET first if need be.
+        try:
+            user_data, created = get_or_create_user_data(email=email, token=token)
+        except NewsletterException as e:
+            return newsletter_exception_response(e)
+
+        update_user.delay(data, user_data['email'], user_data['token'], api_call_type, optin)
```

Comment:
> The sync refers to the fact that we check for user data and send back the correct token, the rest of the update happens async as usual.

Right, I understand that's what's actually happening in the code, but the docstring indicates that the update_user should happen synchronously as well:

>    If sync==True, be sure to do the update synchronously and include the token
>    in the response.  Otherwise, basket has the option to do it all later

If we need to preserve the partially asynchronous behavior, I would suggest a docstring update, possibly combined with changing the kwarg to something like `needs_token`.